### PR TITLE
Master move realm handling to tasks

### PR DESCRIPTION
**Pull request by M4rtinK**
The first few commits add some infrastructure and the new tasks, while the last commit replaces the old code by using the new tasks & there is also a bunch of tests covering the new code.

### CODE REVIEW

**Comment by M4rtinK** [comment]
@AdamWill Hi! :) I'm afraid we don't really have anything at hand to test realm discovery/join works correctly with this change, but IIRC OpenQA has something - would it be possible to do a run on this PR, to check I have not inadvertently broken something? 

Thanks in advance! :)

And let me know if something (updates image ?) else is needed to test this. :)

**Comment by AdamWill** [comment]
Thanks for the headsup, I should be able to do it...I'm working on something else ATM, though, I'll try it later today if I can.

**Comment by M4rtinK** [comment]
> Thanks for the headsup, I should be able to do it...I'm working on something else ATM, though, I'll try it later today if I can.

Thanks! :)

**Comment by M4rtinK** [comment]
Updated with a few fixes & pylint fixes.

File: pyanaconda/installation.py [hunk 1]
```diff
@@ -327,6 +325,23 @@ def run_pre_install():
...
+        modules_with_package_requirements = [SECURITY]
+        for module in modules_with_package_requirements:
+            module_proxy = module.get_proxy()
+            module_requirements = Requirement.from_structure_list(module_proxy.CollectRequirements())
+            for requirement in module_requirements:
+                # check requirement type and add a payload requirement appropriately
+                if requirement.type == "package":
+                    payload.requirements.add_packages([requirement.name], reason=requirement.reason)
+                elif requirement.type == "group":
+                    payload.requirements.add_groups([requirement.name], reason=requirement.reason)
```
Comment (poncovka) [inline pyanaconda/installation.py]: I recommend to implement a new method `add_requirements(requirements)` in the class `PayloadRequirements` and move this code there.

File: pyanaconda/modules/security/security.py [hunk 2]
```diff
@@ -176,15 +180,64 @@ def set_realm(self, realm):
...
         :returns: list of object paths of installation tasks
         """
-        tasks = [
-            ConfigureSELinuxTask(sysroot=conf.target.system_root, selinux_mode=self.selinux)
-        ]
+        tasks = [ConfigureSELinuxTask(sysroot=conf.target.system_root, selinux_mode=self.selinux),
+                 RealmJoinTask(sysroot=conf.target.system_root,
+                               realm_discovered=self.realm.realm_discovered,
+                               realm_name=self.realm.name,
+                               join_options=self.realm.join_options)]
```
Comment (poncovka) [inline pyanaconda/modules/security/security.py]: It looks like `RealmJoinTask` could get just the system root and  `self.realm`.

File: pyanaconda/modules/security/security.py [hunk 3]
```diff
@@ -176,15 +180,64 @@ def set_realm(self, realm):
...
+        for name in self._realm_required_packages:
+            requirements.append(Requirement.for_package(name, reason="realm join"))
+
+        return requirements
+
+    def setup_with_tasks(self):
+        """Return the setup tasks for this module."""
+        realm_task = RealmDiscoverTask(sysroot=conf.target.system_root,
+                                      realm_name=self.realm.name,
+                                      discover_options=self.realm.discover_options)
```
Comment (poncovka) [inline pyanaconda/modules/security/security.py]: It looks like `RealmDiscoverTask` could get just the system root and `self.realm`.

File: pyanaconda/modules/common/structures/realm.py [hunk 4]
```diff
@@ -71,3 +72,12 @@ def join_options(self) -> List[Str]:
     @join_options.setter
     def join_options(self, options: List[Str]):
         self._join_options = options
+
+    @property
+    def realm_discovered(self) -> Bool:
```
Comment (poncovka) [inline pyanaconda/modules/common/structures/realm.py]: I would rename to  `discovered`. It is clear, that it is related to a realm.

File: pyanaconda/modules/security/security.py [hunk 5]
```diff
@@ -52,6 +54,8 @@ def __init__(self):
         self.realm_changed = Signal()
         self._realm = RealmData()
 
+        self._realm_required_packages = []
```
Comment (poncovka) [inline pyanaconda/modules/security/security.py]: I would keep packages in the realm data as well. Please, create a new `packages` property.

File: pyanaconda/modules/security/security.py [hunk 6]
```diff
@@ -176,15 +180,64 @@ def set_realm(self, realm):
...
+
+        return requirements
+
+    def setup_with_tasks(self):
+        """Return the setup tasks for this module."""
+        realm_task = RealmDiscoverTask(sysroot=conf.target.system_root,
+                                      realm_name=self.realm.name,
+                                      discover_options=self.realm.discover_options)
+
+        realm_task.succeeded_signal.connect(lambda: self.handle_realm_discover_results(realm_task.get_result()))
```
Comment (poncovka) [inline pyanaconda/modules/security/security.py]: In this case, I would just update the data structure for realm in the task. You don't need anything else for keeping the results.

File: pyanaconda/modules/common/base/base_interface.py [hunk 7]
```diff
@@ -123,6 +123,14 @@ def CollectRequirements(self) -> List[Structure]:
         """
         return Requirement.to_structure_list(self.implementation.collect_requirements())
 
+    def SetupWithTasks(self) -> List[ObjPath]:
+        """Returns setup tasks of this module.
```
[the remark this answers is not in the document]
Comment (M4rtinK) [reply pyanaconda/modules/common/base/base_interface.py]: Yeah, I'll just use a specific temporary API & we can do this properly once we got task introspection going.

File: pyanaconda/modules/security/security.py (hunk 3 above)
Comment (M4rtinK) [reply pyanaconda/modules/security/security.py]: Yep, this should work out of the box.

File: pyanaconda/modules/security/security.py (hunk 2 above)
Comment (M4rtinK) [reply pyanaconda/modules/security/security.py]: Yep, this should work as long as we update the realm data when realm_changed signal is triggered. :)

**Comment by M4rtinK** [comment]
Updated the PR with a bunch of feedback to @poncovka suggestions & even more tests. :)

File: pyanaconda/modules/security/security.py (hunk 5 above)
Comment (M4rtinK) [reply pyanaconda/modules/security/security.py]: Good point, added. :)

File: pyanaconda/modules/common/structures/realm.py (hunk 4 above)
Comment (M4rtinK) [reply pyanaconda/modules/common/structures/realm.py]: Yep, renamed. :)

File: pyanaconda/installation.py (hunk 1 above)
Comment (M4rtinK) [reply pyanaconda/installation.py]: Good point! Added. :)

**Comment by AdamWill** [comment]
Sorry, I'm having trouble testing this via openQA as Rawhide is too broken ATM :( selinux issues are a problem, and freeipa-server is not installable due to changes in systemd. I'll try and get it cleaned up...

File: pyanaconda/payload/requirement.py [hunk 8]
```diff
@@ -118,6 +118,20 @@ def add_groups(self, group_ids, reason, strong=True):
...
+        """Add requirements from a list of Requirement instances.
+
+        :param requirements: list of Requirement instances
+        """
+        for requirement in requirements:
+            # check requirement type and add a payload requirement appropriately
+            if requirement.type == "package":
+                self.add_packages([requirement.name], reason=requirement.reason)
+            elif requirement.type == "group":
+                self.add_groups([requirement.name], reason=requirement.reason)
```
Comment (poncovka) [inline pyanaconda/payload/requirement.py]: You can use the enum `PayloadRequirementType` and the method `_add` to simplify this code if you want to. However, this implementation is good enough for now.

Comment (M4rtinK) [reply pyanaconda/payload/requirement.py]: I think its fine for now. :)

File: pyanaconda/modules/security/security.py [hunk 9]
```diff
@@ -176,15 +179,51 @@ def set_realm(self, realm):
...
+
+        return requirements
+
+    def discover_realm_with_task(self):
+        """Return the setup task for discovering a realm."""
+        # We do a deep copy of the realm data holder to avoid
+        # changes to the data in the Task from changing the backing
+        # data structure in the module, without triggering the changed signal.
+        realm_task = RealmDiscoverTask(sysroot=conf.target.system_root,
+                                      realm_data=copy.deepcopy(self.realm))
```
[the remark this answers is not in the document]
Comment (M4rtinK) [reply pyanaconda/modules/security/security.py]: Well, this is also about the realm data potentially changing after the `RealmDiscoverTask` has been instantiated. By doing a deepcopy outright we avoid both data changes comming by reference from the module to task and the other way around.

File: pyanaconda/modules/security/security.py [hunk 10]
```diff
@@ -176,15 +179,51 @@ def set_realm(self, realm):
...
+    def collect_requirements(self):
+        """Return installation requirements for this module.
+
+        :return: a list of requirements
+        """
+        requirements = []
+
+        # Add realm requirements.
+        for name in self.realm.required_packages:
+            requirements.append(Requirement.for_package(name, reason="needed to join a realm"))
```
[the remark this answers is not in the document]
Comment (M4rtinK) [reply pyanaconda/modules/security/security.py]: I think a more descriptive reason is good & should be more widely used.

The main use case for the reason field is people reading logs to find out why some packages have been automatically added to their system, so it the reson should IMHO be easy to read & descriptive.

**Comment by AdamWill** [comment]
openQA test came up [no good](https://openqa.stg.fedoraproject.org/tests/582808). Problem seems to be this:

    var/log/anaconda/syslog:04:20:34,861 WARNING org.fedoraproject.Anaconda.Modules.Security:INFO:program:realm: Couldn't join realm: This computer's host name is not set correctly.
    var/log/anaconda/syslog:04:20:34,862 WARNING org.fedoraproject.Anaconda.Modules.Security:INFO:anaconda.modules.security.installation:Joining realm ipa001.domain.local failed

File: pyanaconda/modules/security/security.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (poncovka) [reply pyanaconda/modules/security/security.py]: You can create the copy in the `__init__` method of the task.

File: pyanaconda/installation.py [hunk 11]
```diff
@@ -167,11 +168,6 @@ def _prepare_configuration(storage, payload, ksdata):
 
     configuration_queue.append(generate_initramfs)
 
-    # join a realm (if required)
-    join_realm = TaskQueue("Realm join", N_("Joining realm"))
```
Comment (poncovka) [inline pyanaconda/installation.py]: The task for joining the realm needs to run after setting up the network, but the tasks of the Security module are running before that. I guess that the best solution is to create a new DBus method for this task for now. Please, document this requirement in the docstring of `DiscoverRealmWithTask`.

File: pyanaconda/modules/security/security.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (poncovka) [reply pyanaconda/modules/security/security.py]: The use case you are talking about can be solved on the module level. You can easily log why you add the specific packages and groups to the list of requirements. I don't understand why we should propagate this detailed information on DBus.

I prefer to use just a name of a component or a tool that requires it (as it was before), so the requirements can be easily processed by a third party. For example, when you want to remove all requirements for `realm`.

We should discuss how the attribute `reason` should look like. One of the solutions could be to extend the DBus structure for the requirement to be able to specify the name of the component and the detailed description.

[the remark this answers is not in the document]
Comment (M4rtinK) [reply pyanaconda/modules/security/security.py]: > We should discuss how the attribute `reason` should look like. One of the solutions could be to extend the DBus structure for the requirement to be able to specify the name of the component and the detailed description.

This sounds like an ideal solution to me.

File: pyanaconda/modules/security/security.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (M4rtinK) [reply pyanaconda/modules/security/security.py]: Good point, that indeed looks nice! :) Changed.

File: pyanaconda/installation.py (hunk 11 above)
Comment (M4rtinK) [reply pyanaconda/installation.py]: Indeed! I've split out the task geter to a separate & temporary DBUS method and moved its invocation to the original place where the kickstart execute method has been previously. That should hopefully fix things. :)

**Comment by M4rtinK** [comment]
> openQA kickstart join test came up [no good](https://openqa.stg.fedoraproject.org/tests/582808). Problem seems to be this:
> 
> ```
> var/log/anaconda/syslog:04:20:34,861 WARNING org.fedoraproject.Anaconda.Modules.Security:INFO:program:realm: Couldn't join realm: This computer's host name is not set correctly.
> var/log/anaconda/syslog:04:20:34,862 WARNING org.fedoraproject.Anaconda.Modules.Security:INFO:anaconda.modules.security.installation:Joining realm ipa001.domain.local failed
> ```

@AdamWill Thanks! Should be hopefully fixed now, so please retest with the updated PR. :)

**Comment by AdamWill** [comment]
OK, will do.

**Comment by AdamWill** [comment]
sorry, some other shiny things appeared in front of me first...will try it again soon.

**Comment by AdamWill** [comment]
openQA test with current version of the PR [passed](https://openqa.stg.fedoraproject.org/tests/58432), yay.

**Comment by M4rtinK** [comment]
@AdamWill Nice! Thanks a lot! :)

File: pyanaconda/modules/security/installation.py [hunk 12]
```diff
@@ -72,3 +77,119 @@ def run(self):
...
+            for line in lines:
+                parts = line.split(":", 1)
+                if len(parts) == 2 and parts[0].strip() == "required-package":
+                    package_spec = parts[1].strip()
+                    # "" is not a valid package specification
+                    if package_spec:
+                        required_packages.append(package_spec)
+
+            log.info("Realm %s needs packages %s",
+                      discovered_realm_data, ", ".join(required_packages))
```
[the remark this answers is not in the document]
Comment (M4rtinK) [reply pyanaconda/modules/security/installation.py]: Good point, split the parsing off. :)

File: tests/nosetests/pyanaconda_tests/module_security_test.py [hunk 13]
```diff
@@ -154,3 +165,598 @@ def realm_kickstart_test(self):
...
+
+        publisher.assert_called()
+
+        # realm discover
+        object_path = publisher.call_args_list[0][0][0]
+        obj = publisher.call_args_list[0][0][1]
+
+        self.assertEqual(realm_discover_task_path, object_path)
+        self.assertIsInstance(obj, TaskInterface)
+        self.assertIsInstance(obj.implementation, RealmDiscoverTask)
```
[the remark this answers is not in the document]
Comment (M4rtinK) [reply tests/nosetests/pyanaconda_tests/module_security_test.py]: Yep, this is indeed handy for the `*WithTask()` methods that return a single task, but I think we will need to extend this wot cover also the `*WithTasks()` methods that return a list of task paths.

**Comment by M4rtinK** [comment]
jenkins, test this please

File: tests/nosetests/pyanaconda_tests/module_security_test.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (jkonecny12) [reply tests/nosetests/pyanaconda_tests/module_security_test.py]: Yep, it would be handy.

**Pull request merged**